### ISSUE TITLE
perf: TensorWorkspace + JIT compiler integration to surpass PyTorch performance

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#38]
## Goal

Make AiDotNet.Tensors + AiDotNet JIT compiler **faster than PyTorch** across all common operations. This requires completing TensorWorkspace, integrating it with the JIT compiler, adding missing optimization passes, and proving superiority with benchmarks.

### What We Have

- **TensorWorkspace** (PR #37): Pre-allocates a single contiguous buffer for all intermediate tensors. Returns pre-mapped Tensor views via pointer arithmetic (0ns vs PyTorch ~10ns caching allocator). But it is a dumb memory pool with no JIT integration, no lifetime analysis, no graph compilation.

- **JIT Compiler** (AiDotNet): 8 optimization passes, 170+ IR operations, 20+ fused operations, GPU code generation, SIMD optimizer. But it is not wired into TensorWorkspace or the forward pass of any model.

- **In-Place IEngine Ops** (PR #37): TensorAddInPlace, TensorBroadcastAddInPlace, Conv2DInto, SigmoidInPlace, ReLUInPlace, GroupNormInto with all GPU backends. Eliminates per-operation allocation.

- **FusedConv2D**: Conv + Bias + Activation in single kernel call, already in IEngine with all GPU backends.

- **10 Fusion Patterns**: MatMul+Add+Activation, Conv2D+BatchNorm, Conv2D+Add(bias), Residual+Activation, BatchNorm+Activation, LayerNorm+Add, element-wise chains.

#### Memory Planning (Critical)

- [x] **Tensor lifetime analysis**: Compute when each intermediate tensor is last used, then reuse its memory for subsequent tensors. PyTorch TorchInductor does this to minimize peak memory. Our TensorWorkspace allocates a separate slot for every tensor with no reuse.

- [x] **In-place operation scheduling**: When a tensor last consumer can write its output in-place (overwriting the input), schedule it that way. Eliminates the intermediate entirely.

- [x] **Memory pool with size classes**: For dynamic shapes (varying batch sizes), maintain pools of common sizes. Return to pool on last use. Surpass PyTorch caching allocator by using size-exact pools (no power-of-2 rounding waste).

#### JIT Compilation Integration (Critical)

- [x] **TensorWorkspace as JIT memory backend**: The JIT compiler code generator should emit code that reads/writes workspace slots instead of allocating tensors. Each IR node maps to a workspace slot ID.

- [x] **Graph capture mode**: Add a mode where the first forward pass records all operations and tensor shapes into a computation graph, then JIT-compiles it. Subsequent forward passes execute the compiled graph. (Similar to torch.compile / tf.function / JAX jit.)

- [x] **Compiled forward pass for UNet**: The UNet ForwardUNet method should have a JIT-compiled path that executes ALL layers as a single compiled function with no Python-like layer-by-layer interpretation.

- [x] **Cache compiled graphs**: Key by input shape + model architecture hash. Reuse across inference calls.

#### Missing Optimization Passes (High Priority)

- [x] **GroupNorm + Activation fusion**: DiffusionResBlock does GroupNorm then SiLU repeatedly. Fusing these eliminates one of the two largest intermediate tensors per ResBlock.

- [x] **Conv2D + Bias + Activation fusion IR op**: We have FusedConvBatchNormActivationOp but need Conv2D+Bias+Activation without BatchNorm (the diffusion UNet pattern). The IEngine FusedConv2D exists but has no IR op.

- [x] **Tile scheduling pass**: Partition large operations into tiles that fit in L1/L2 cache. Critical for Conv2D and MatMul performance on CPU.

- [x] **Operator reordering pass**: Reorder independent operations to maximize data locality. When tensor A is consumed by ops B and C, schedule whichever one shares more cache lines with A first.

- [x] **Buffer aliasing pass**: After lifetime analysis, assign multiple dead tensors to the same workspace slot. This is the key to matching/beating PyTorch memory efficiency.

- [x] **Quantization-aware fusion**: INT8/FP16 fused kernels for inference. PyTorch has torch.quantization; we need equivalent capability.

- [x] **Dynamic shape support**: Compile once with symbolic shapes, specialize at runtime. Avoid recompilation when only batch size changes.

- [x] **Graph partitioning**: Split computation graph into CPU and GPU subgraphs based on operation support and data size. Execute each on optimal hardware.

#### Missing IR Operations (Medium Priority)

- [x] **GroupNormOp / FusedGroupNormActivationOp**: GroupNorm is used in every DiffusionResBlock but has no IR representation. Cannot fuse GroupNorm+SiLU without it.

- [x] **FusedConv2DBiasActivationOp**: Maps to IEngine.FusedConv2D. Different from FusedConvBatchNormActivationOp (no batch norm).

- [x] **ScatterNdOp / GatherNdOp**: For sparse and attention operations.

- [x] **FlashAttentionOp**: Fused multi-head attention that never materializes the NxN attention matrix.

- [x] **FusedResidualGroupNormOp**: For the DiffusionResBlock pattern: h = GroupNorm(conv_output + residual). Fuses the add + norm.

#### Benchmarks (Critical - Cannot Prove Anything Without Numbers)

- [x] **BenchmarkDotNet project**: Create benchmarks/AiDotNet.Tensors.Benchmarks/ with BenchmarkDotNet

- [x] **Operation-level benchmarks**: Conv2D, MatMul, GroupNorm, Attention at various sizes. Compare against PyTorch (via TorchSharp) and native BLAS.

- [x] **Layer-level benchmarks**: ConvolutionalLayer, DiffusionResBlock, MultiHeadAttention forward pass. Compare JIT-compiled vs interpreted vs PyTorch.

- [x] **Model-level benchmarks**: UNet forward pass (SD15 dimensions), full Predict (50 denoising steps). Compare against PyTorch Stable Diffusion.

- [x] **Memory benchmarks**: Peak memory usage, allocation count, GC pressure. Compare TensorWorkspace vs PyTorch caching allocator.

- [x] **Publish results**: Markdown table in README showing operation/layer/model performance vs PyTorch with percentage improvement.

#### SIMD / Vectorization Gaps (Medium Priority)

- [x] **Double-precision SIMD Exp/Log/Tanh**: The double path uses scalar Math functions while float has full AVX2 SIMD. Add Vector256 double versions.

- [x] **Vectorized RNG**: Weight initialization uses scalar Random.NextDouble(). Use SIMD-parallel RNG (e.g., xoshiro256** with AVX2) for 4-8x faster initialization.

- [x] **NHWC layout support**: Channels-last memory layout enables better vectorization for Conv2D on CPU (contiguous channel access). PyTorch supports both NCHW and NHWC.

#### GPU Optimization (Lower Priority - CPU First)

- [x] **GPU graph execution**: Submit entire compiled graph as a single GPU command buffer instead of per-operation kernel launches. Eliminates kernel launch overhead.

- [x] **Persistent kernels**: For operations that reuse the same data (attention), keep data in GPU registers/shared memory across kernel boundaries.

- [x] **Multi-stream execution**: Overlap independent operations on different GPU streams.

- [x] **GPU memory workspace**: TensorWorkspace backed by a single GPU allocation, with sub-allocations for each slot. Zero GPU-side allocation during inference.

## Architecture

```



User Code (UNet.Forward)



    |



    v



Graph Capture Mode (first call records operations)



    |



    v



JIT Compiler Pipeline:



    IR Builder -> IR Graph



    -> Constant Folding Pass



    -> Dead Code Elimination Pass



    -> Operation Fusion Pass (10+ patterns including GroupNorm+SiLU, Conv+Bias+Act)



    -> Buffer Aliasing Pass (lifetime analysis -> minimum memory)



    -> Tile Scheduling Pass (cache-optimal blocking)



    -> Operator Reordering Pass (data locality)



    -> Vectorization Pass (SIMD width selection)



    -> Auto-Tuning Pass (runtime profiling)



    -> Code Generator (CPU SIMD / GPU kernels)



    |



    v



Compiled Function (backed by TensorWorkspace memory)



    |



    v



Engine.Execute(compiledFunction, inputTensors)



    -> Zero allocation, fused kernels, optimal memory layout



```

## Success Criteria

1. **Conv2D (1280x1280x3x3 on 8x8 spatial, double)**: Faster than PyTorch CPU

2. **DiffusionResBlock forward**: Faster than PyTorch ResBlock with same architecture

3. **UNet forward pass (SD15 default dimensions)**: Faster than PyTorch UNet

4. **50-step denoising (SD15 Predict)**: Faster than PyTorch Stable Diffusion pipeline

5. **Peak memory during 50-step denoising**: Lower than PyTorch

6. **GC pauses during inference**: Zero (all memory pre-allocated)

7. **All benchmarks published** with reproducible numbers

## Quick Wins for Diffusion (Do First)

These can be done without the full JIT integration:

1. GroupNormOp IR + FusedGroupNormActivationOp for GroupNorm+SiLU fusion

2. FusedConv2DBiasActivationOp for Conv+Bias+SiLU without going through 3 separate calls

3. Memory planning pass to compute minimum workspace size for UNet

4. Benchmark infrastructure (BenchmarkDotNet project with Conv2D and ResBlock benchmarks)

## Related PRs

- #37: In-place IEngine ops + TensorWorkspace foundation

- #36: Double precision Conv2D im2col+GEMM

## References

- PyTorch TorchInductor: https://pytorch.org/docs/stable/torch.compiler.html

- TVM: An Automated End-to-End Optimizing Compiler: https://arxiv.org/abs/1802.04799

- Triton: An Intermediate Language for Block-Level GPU Programming

- XLA: Optimizing Compiler for Machine Learning: https://www.tensorflow.org/xla
